compiler/code-gen/declarations.cpp

```diff
@@ -596,27 +619,120 @@ void ClassDeclaration::compile_get_hash(CodeGenerator &W, ClassPtr klass) {
   compile_class_method(FunctionSignatureGenerator(W).set_const_this(), klass, "int get_hash()", klass->get_hash());
 }
 
-void ClassDeclaration::compile_accept_visitor(CodeGenerator &W, ClassPtr klass, const char *visitor_type) {
+void ClassDeclaration::compile_accept_visitor(CodeGenerator &W, ClassPtr klass, std::string_view visitor_type) {
   compile_class_method(FunctionSignatureGenerator(W), klass, fmt_format("void accept({} &visitor)", visitor_type), "generic_accept(visitor)");
 }
 
+static void compile_generic_accept(CodeGenerator &W, ClassPtr klass) noexcept {
+  FunctionSignatureGenerator(W) << "template<class Visitor>" << NL
+                                << "void generic_accept(Visitor &&visitor)" << BEGIN;
+
+  for (auto cur_klass = klass; cur_klass; cur_klass = cur_klass->parent_class) {
+    cur_klass->members.for_each([&W](const ClassMemberInstanceField &field) {
+      W << "visitor(\"" << field.local_name() << "\", $" << field.local_name() << ");" << NL;
+    });
+  }
+
+  W << END << NL;
+}
+
+static auto compile_skip_if_default(CodeGenerator &W,
+  const ClassMemberInstanceField &field, bool write_visitor, bool skip_if_default) noexcept {
+  auto restore_action = vk::finally([&W] { W << END << NL; });
+  if (write_visitor && skip_if_default) {
+    W << "if (";
+    if (field.var->init_val) {
+      W << "!eq2($" << field.local_name() << ", " << field.var->init_val << ")";
+    } else {
+      W << "f$boolval($" << field.local_name() << ")";
+    }
+    W << ") " << BEGIN;
+  } else {
+    restore_action.disable();
+  }
+  return restore_action;
+}
+
+static auto compile_set_precision(CodeGenerator &W, bool write_visitor, int precision) noexcept {
+  auto restore_action = vk::finally([&W] { W << "visitor.restore_precision();" << NL; });
+  if (write_visitor && precision) {
+    W << "visitor.set_precision(" << precision << ");" << NL;
+  } else {
+    restore_action.disable();
+  }
+  return restore_action;
+}
+
+static std::string compile_json_field_name_value(CodeGenerator &W, bool json_raw_string_field, vk::string_view local_name) noexcept {
+  if (json_raw_string_field) {
+    W << "JsonRawString __prefix__$" << local_name << "{$" << local_name << "};" << NL;
+    return "__prefix__$" + local_name;
+  } else {
+    return "$" + local_name;
+  }
+}
+
+static vk::string_view get_json_field_name(const kphp_json::FieldTags &tags, vk::string_view local_name) noexcept {
+  auto rename = tags.get_rename_value();
+  return rename.empty() ? local_name : rename;
```

Comment:
`@kphp-json   rename =   "a"` — this generated invalid C++ code due to quotes. Probably, this should be checked when parsing phpdoc. (trim quotes?)

Reply:
I forbid quotes in this tag